<a href="https://colab.research.google.com/github/carlos-alves-one/-Energy-Forecast/blob/main/datasets_merge_V3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Mount Google Drive

In [1]:
# Imports the 'drive' module from 'google.colab' and mounts the Google Drive to
# the '/content/drive' directory in the Colab environment.
from google.colab import drive

# This function mounts Google Drive
def mount_google_drive():
    drive.mount('/content/drive')

# Call the function to mount Google Drive
mount_google_drive()


Mounted at /content/drive


# Importing Necessary Libraries and Packages

In [2]:
import pandas as pd              # Import pandas for data manipulation and analysis


# Input Data Files

In [12]:
# Initial datasets to be merged
test        = pd.read_csv('/content/drive/MyDrive/project_energy/test.csv')
targets     = pd.read_csv('/content/drive/MyDrive/project_energy/revealed_targets.csv')
client      = pd.read_csv('/content/drive/MyDrive/project_energy/client.csv')
historical  = pd.read_csv('/content/drive/MyDrive/project_energy/historical_weather.csv')
forecast    = pd.read_csv('/content/drive/MyDrive/project_energy/forecast_weather.csv')
electricity = pd.read_csv('/content/drive/MyDrive/project_energy/electricity_prices.csv')
gas         = pd.read_csv('/content/drive/MyDrive/project_energy/gas_prices.csv')

# Final dataset with all datasets merged
data        = pd.read_csv('/content/drive/MyDrive/project_energy/data.csv')


# Merge Test and Targets Datasets

The best dataset to start the merge process would be the one that acts as a central hub, having key identifiers that are most common across the other datasets. Based on the critical column presence analysis, the **`test`** dataset is a suitable starting point for merging. Here is why:

1. **Universal Identifier**: The `test` dataset contains the `data_block_id` column, which is present in all other datasets. This makes it an excellent candidate for initial merging, as `data_block_id` can serve as a primary key to link data across different datasets.

2. **Additional Common Identifiers**: It also includes `county`, `product_type`, and `is_business`, which are present in several other datasets (revealed_targets and client). These columns further facilitate merging and provide additional layers of information.

3. **Role in Analysis**: The `test` dataset likely represents the primary data structure into which other data (like weather, prices, and client information) will be integrated. This makes it a logical starting point for building a comprehensive dataset.

Starting with the `test` dataset, we can incrementally merge other datasets like `revealed_targets`, `client`, and then bring in weather and price data, ensuring alignment of time-related variables (like `datetime`, `forecast_date`, `origin_date`) and geographical data (like `latitude` and `longitude`) where applicable.

In [13]:
# Merging the test and revealed_targets datasets on the common columns
merged_df = pd.merge(test, targets,
                     on=['county', 'is_business', 'product_type', 'data_block_id'],
                     how='inner')

# Renaming columns to differentiate between the datasets
merged_df = merged_df.rename(columns={'datetime': 'target_datetime', 'target': 'actual_target'})

# Displaying the merged dataframe
merged_df.head(3).T


0                    1  \
county                                  0                    0   
is_business                             0                    0   
product_type                            1                    1   
is_consumption_x                        0                    0   
prediction_datetime   2023-05-28 00:00:00  2023-05-28 00:00:00   
data_block_id                         634                  634   
row_id_x                          2005872              2005872   
prediction_unit_id_x                    0                    0   
currently_scored                    False                False   
actual_target                       2.675              471.887   
is_consumption_y                        0                    1   
target_datetime       2023-05-26 00:00:00  2023-05-26 00:00:00   
row_id_y                          1999536              1999537   
prediction_unit_id_y                    0                    0   

                                        2  
county                                  0  
is_business                             0  
product_type                            1  
is_consumption_x                        0  
prediction_datetime   2023-05-28 00:00:00  
data_block_id                         634  
row_id_x                          2005872  
prediction_unit_id_x                    0  
currently_scored                    False  
actual_target                       2.138  
is_consumption_y                        0  
target_datetime       2023-05-26 01:00:00  
row_id_y                          1999668  
prediction_unit_id_y                    0

In [14]:
test.shape

(12480, 9)

In [15]:
targets.shape

(12576, 9)

In [16]:
merged_df.shape

(599040, 14)

In [17]:
print(merged_df.isnull().sum())

county                  0
is_business             0
product_type            0
is_consumption_x        0
prediction_datetime     0
data_block_id           0
row_id_x                0
prediction_unit_id_x    0
currently_scored        0
actual_target           0
is_consumption_y        0
target_datetime         0
row_id_y                0
prediction_unit_id_y    0
dtype: int64


In [18]:
# Saving the final merged dataset to CSV
output_file_path = "/content/drive/MyDrive/project_energy/merged_data.csv"
merged_df.to_csv(output_file_path, index=False)  # Set index=False if you don't want to include the index

In [19]:
# Load the merged dataset and display the first 3 records
data = pd.read_csv("/content/drive/MyDrive/project_energy/merged_data.csv")
data.head(3).T

0                    1  \
county                                  0                    0   
is_business                             0                    0   
product_type                            1                    1   
is_consumption_x                        0                    0   
prediction_datetime   2023-05-28 00:00:00  2023-05-28 00:00:00   
data_block_id                         634                  634   
row_id_x                          2005872              2005872   
prediction_unit_id_x                    0                    0   
currently_scored                    False                False   
actual_target                       2.675              471.887   
is_consumption_y                        0                    1   
target_datetime       2023-05-26 00:00:00  2023-05-26 00:00:00   
row_id_y                          1999536              1999537   
prediction_unit_id_y                    0                    0   

                                        2  
county                                  0  
is_business                             0  
product_type                            1  
is_consumption_x                        0  
prediction_datetime   2023-05-28 00:00:00  
data_block_id                         634  
row_id_x                          2005872  
prediction_unit_id_x                    0  
currently_scored                    False  
actual_target                       2.138  
is_consumption_y                        0  
target_datetime       2023-05-26 01:00:00  
row_id_y                          1999668  
prediction_unit_id_y                    0

### Dataset Overview
- **Total Entries:** 599,040
- **Columns:** 14

### Column Details
1. **county:** Integer values, ranging from 0 to 15.
2. **is_business:** Binary integer (0 or 1), indicating if the entry pertains to a business.
3. **product_type:** Integer values, ranging from 0 to 3, possibly categorizing types of products or services.
4. **is_consumption_x and is_consumption_y:** Binary integers, likely indicating consumption status. Both columns are identical in terms of value distribution.
5. **prediction_datetime and target_datetime:** Date-time strings, likely representing different timestamps for prediction and target data.
6. **data_block_id:** Integer values, ranging from 634 to 637, potentially signifying different data batches or segments.
7. **row_id_x and row_id_y:** Integer identifiers for rows with a wide range of values.
8. **prediction_unit_id_x and prediction_unit_id_y:** Integer values, suggesting some form of categorization or grouping for predictions.
9. **currently_scored:** Boolean values, possibly indicating whether a prediction is being evaluated.
10. **actual_target:** Floating-point numbers representing a target metric with a broad range and substantial variance.

### Statistical Summary
- The dataset shows a wide variance in the `actual_target` column, with values ranging from 0 to over 11,000.
- The mean `actual_target` is around 388, but the median is significantly lower at about 64.64, indicating a right-skewed distribution.
- The columns `county`, `is_business`, `product_type`, and `is_consumption_x/y` show a uniform distribution across their respective ranges.

### Data Quality
- No missing values in any of the columns indicate good data completeness.

### Implications and Potential Analysis
- **Prosumer Behavior:** By examining the `actual_target` with `product_type`, `county`, and `is_business`, we can infer energy consumption or production patterns.
- **Temporal Analysis:** With `prediction_datetime` and `target_datetime`, it is possible to conduct time series analysis to understand trends and seasonal variations.
- **Business vs. Individual Analysis:** The `is_business` column can help differentiate consumption or production patterns between businesses and individuals.

This dataset is well-suited for detailed energy consumption and production patterns analysis, especially considering prosumer behaviour in the energy sector. Further analysis could involve exploring correlations between variables and building predictive models for energy demand or supply.

# Merge Data and Client Datasets


In [20]:
# Merging the previously merged dataset with the client dataset
merged_with_client_df = pd.merge(data, client,
                                 on=['county', 'is_business', 'product_type'],
                                 how='inner')

# Display the first few rows of the newly merged dataset
merged_with_client_df.head(3).T


0                    1  \
county                                  0                    0   
is_business                             0                    0   
product_type                            1                    1   
is_consumption_x                        0                    0   
prediction_datetime   2023-05-28 00:00:00  2023-05-28 00:00:00   
data_block_id_x                       634                  634   
row_id_x                          2005872              2005872   
prediction_unit_id_x                    0                    0   
currently_scored                    False                False   
actual_target                       2.675                2.675   
is_consumption_y                        0                    0   
target_datetime       2023-05-26 00:00:00  2023-05-26 00:00:00   
row_id_y                          1999536              1999536   
prediction_unit_id_y                    0                    0   
eic_count                             507                  507   
installed_capacity               4960.215             4960.215   
date                           2023-05-26           2023-05-27   
data_block_id_y                       634                  635   

                                        2  
county                                  0  
is_business                             0  
product_type                            1  
is_consumption_x                        0  
prediction_datetime   2023-05-28 00:00:00  
data_block_id_x                       634  
row_id_x                          2005872  
prediction_unit_id_x                    0  
currently_scored                    False  
actual_target                       2.675  
is_consumption_y                        0  
target_datetime       2023-05-26 00:00:00  
row_id_y                          1999536  
prediction_unit_id_y                    0  
eic_count                             508  
installed_capacity               4964.215  
date                           2023-05-28  
data_block_id_y                       636

The merged dataset, which now includes data from the test, revealed_targets, and client datasets, has been successfully created. This new dataset combines a wide range of information, including county, business status, product type, consumption patterns, prediction details, and client-specific data like eic_count (possibly referring to the number of clients or entities) and installed_capacity.

In [21]:
merged_with_client_df.shape

(2396160, 18)

In [22]:
client.shape

(262, 7)

In [23]:
print(merged_with_client_df.isnull().sum())

county                  0
is_business             0
product_type            0
is_consumption_x        0
prediction_datetime     0
data_block_id_x         0
row_id_x                0
prediction_unit_id_x    0
currently_scored        0
actual_target           0
is_consumption_y        0
target_datetime         0
row_id_y                0
prediction_unit_id_y    0
eic_count               0
installed_capacity      0
date                    0
data_block_id_y         0
dtype: int64


In [26]:
# Saving the final merged dataset to CSV
output_file_path = "/content/drive/MyDrive/project_energy/merged_data.csv"
merged_with_client_df.to_csv(output_file_path, index=False)  # Set index=False if you don't want to include the index


In [27]:
# Load the merged dataset and display the first 3 records
data = pd.read_csv("/content/drive/MyDrive/project_energy/merged_data.csv")
data.head(3).T


0                    1  \
county                                  0                    0   
is_business                             0                    0   
product_type                            1                    1   
is_consumption_x                        0                    0   
prediction_datetime   2023-05-28 00:00:00  2023-05-28 00:00:00   
data_block_id_x                       634                  634   
row_id_x                          2005872              2005872   
prediction_unit_id_x                    0                    0   
currently_scored                    False                False   
actual_target                       2.675                2.675   
is_consumption_y                        0                    0   
target_datetime       2023-05-26 00:00:00  2023-05-26 00:00:00   
row_id_y                          1999536              1999536   
prediction_unit_id_y                    0                    0   
eic_count                             507                  507   
installed_capacity               4960.215             4960.215   
date                           2023-05-26           2023-05-27   
data_block_id_y                       634                  635   

                                        2  
county                                  0  
is_business                             0  
product_type                            1  
is_consumption_x                        0  
prediction_datetime   2023-05-28 00:00:00  
data_block_id_x                       634  
row_id_x                          2005872  
prediction_unit_id_x                    0  
currently_scored                    False  
actual_target                       2.675  
is_consumption_y                        0  
target_datetime       2023-05-26 00:00:00  
row_id_y                          1999536  
prediction_unit_id_y                    0  
eic_count                             508  
installed_capacity               4964.215  
date                           2023-05-28  
data_block_id_y                       636